### ISSUE TITLE
Calling client.query() *after* calling done()?

### ISSUE THREAD

**Issue by HarryPehkonen** [brianc/node-postgres#771]
Is it not bad to call done() before running the query?  I thought there would be an error:

    var pg = require('pg');
    
    pg.connect(function(err, client, done) {
      if (err) {
        console.log('error while connecting:  ' + err.message);
      } else {
        console.log('connected');
    
        // call done before query
        done();
    
        // call query after done
        client.query('SELECT NOW() AS "theTime"', function(err, result) {
          if (err) {
            console.log('error from query:  ' + err.message);
          } else {
            console.log('result:  ' + JSON.stringify(result, null, 4));
          }
          pg.end();
        });
      }
    });

Specifically:

* Using pg.connect, so I'm getting clients from the pool
* Call done() so the client gets returned to the pool
* Call client.query() right away, or even in nextTick()

**Comment by sehrope** [comment]
Yes it's bad. You shouldn't call `done()` till you are done using the pooled connection.

There won't necessarily be an error as the connection most likely still exists in the pool. In fact, you'll probably never run into an error in a stand alone test like that. The issue comes up if you have a separate event loop (ex: another web request) that gets the connection from the pool and tries to use it at the same time.

Long story short, don't call `done()` till you no longer need the connection. Also, if there is an error using the connection (ex: a connect or query error), you probably want to call `done(err)` to have the connection removed from the pool (rather than just returned to it).

**Comment by brianc** [comment]
Pretty much exactly what @sehrope said. :)
